python/phase-2.md

```diff
@@ -828,4 +829,168 @@ Notes:
+  similar that is smart enough to convert to an activity definition. I have been trying to avoid actually wrapping
+  functions so far.
+* We could support the following with an overload:
+
+    ```python
+    workflow.execute_activity(ActivityClass.some_activity, "foo", method=True)
+    ```
+  This would work with a special overload like:
+
+    ```python
+    @overload
+    async def execute_activity(
+        activity: Callable[[Any, LocalParamType], LocalReturnType],
+        arg: LocalParamType,
+        /,
+        *,
+        method: Literal[True],
+    ```
+  This seems like a fairly sane way, but requires callers pass a special parameter to call
+* Another outstanding question - is there added benefit _at this time_ to support registering a class full of activity
+  methods instead of individually? I am thinking not (yet).
+
+### Workflow Contracts
+
+Sometimes users may want to separate the workflow into a contract and an impl so that they can be separated. We will
+support this approach.
+
+Here is what it looks like with `typing.Protocol`s:
+
+```python
+@workflow.defn
+class MyWorkflow(Protocol):
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        ...
+
+@workflow.defn(name=MyWorkflow.__name__)
+class MyWorkflowImpl:
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        return [arg]
+
+async def run_workflow(client: Client) -> List[str]:
+    return await client.execute_workflow(MyWorkflow.run)
+
+def create_worker(client: Client) -> Worker:
+    return Worker(workflows=[MyWorkflowImpl])
+```
+
+And here's what it looks like with abstract classes:
+
+```python
+@workflow.defn
+class MyWorkflow(ABC):
+  @workflow.run
+  @abstractmethod
+  def run(self, arg: str) -> List[str]:
+    raise NotImplementedError
+
+@workflow.defn(name=MyWorkflow.__name__)
+class MyWorkflowImpl(MyWorkflow):
+    @workflow.run
+    def run(self, arg: str) -> List[str]:
+        return [arg]
+
+async def run_workflow(client: Client) -> List[str]:
+    return await client.execute_workflow(MyWorkflow.run)
+
+def create_worker(client: Client) -> Worker:
+    return Worker(workflows=[MyWorkflowImpl])
+```
+
+Notes:
+
+* This has no special logic to handle this approach. The same approach can be used to separate activity class method
+  contracts from their impls.
+* It is unfortunate that we require a `@workflow.defn` on both the base and impl classes.
+  * We want to require this decorator at client-side _and_ worker-side to ensure name sharing
+  * With `Protocol` it is obvious why, there is technically no code shared
+  * With abstract base classes, we maybe could support `@workflow.defn(extends=MyWorkflow)` but that doesn't give you
+    much
+  * I am hesitant to try to implicitly derive the base name because there are other use cases for inheritance besides
+    just contract/impl separation
+* It is unfortunate that we require a `@workflow.run` on both the base and impl classes.
```

Comment:
All of these complexities with abstract classes make me wonder if it's worth supporting them at all.
Maybe we shouldn't support inheritance if we can support protocols.

Reply:
Or at least not support abstract classes but I'm not convincing myself.

Reply:
> Maybe we shouldn't support inheritance if we can support protocols.

There is no reason not to support inheritance. The only problems you are seeing here are brought on by the fact that you want to separate contract from the implementation . Normal inheritance poses no problems and we need to have no special support for it.